File: packages/go/report/html/html.go
// Package html renders runner results as a single self-contained HTML page
// for human inspection — no external assets, no JavaScript. The look and
// feel is a dark-first compatibility dashboard (light theme via
// prefers-color-scheme): a header with an overall split badge of outcome
// counts, a coverage-style summary table of groups (pass-fraction bars,
// watermarked pass percentages, per-outcome counts), and a
// staged-disclosure section per group whose vector cards open to
// expected-vs-actual detail. Groups with failures start expanded (their
// vector cards stay collapsed until clicked); clean groups start collapsed.
// Navigation is anchors only: group names jump to their section and the
// fail badge jumps to the first failing vector.
//
// The template is mustache, shared verbatim across runner implementations
// (canonical copy: shared/report/page.mustache, synced by
// scripts/sync-report-template.js). Every implementation builds the same
// JSON-shaped view model — sections switch only on booleans and arrays, and
// all numbers are pre-formatted with integer arithmetic — so identical
// results render byte-identical pages in every language, enforced by the
// shared golden test (shared/report/fixture.json + golden.html).
//
// The headline pass rate is pass / (total − skipped): everything the runner
// attempted counts in the denominator (blocked vectors and runner errors
// included — their counts are displayed separately so the nuance stays
// visible).
//
// Output is byte-deterministic for identical input: the generation time is
// never sampled by the reporter itself — set report.Meta.GeneratedAt to
// stamp one (shown in the provenance panel and footer, in UTC).
package html

import (
	_ "embed"
	"fmt"
	"io"
	"iter"
	"sort"
	"strings"
	"time"

	"github.com/cbroglie/mustache"

	s3tests "github.com/cloud-portable/s3tests/packages/go"
	"github.com/cloud-portable/s3tests/packages/go/report"
)

//go:embed page.mustache
var pageSrc string

var page = func() *mustache.Template {
	t, err := mustache.ParseString(pageSrc)
	if err != nil {
		panic("s3tests: embedded report template invalid: " + err.Error())
	}
	return t
}()

// Write consumes the results and writes the report page.
func Write(w io.Writer, results iter.Seq[s3tests.VectorResult], meta report.Meta) error {
	return page.FRender(w, viewModel(results, meta))
}

// viewModel builds the shared cross-implementation view model: a JSON-shaped
// map whose conditionals are explicit booleans and whose numbers are
// pre-formatted strings (see the package comment).
func viewModel(results iter.Seq[s3tests.VectorResult], meta report.Meta) map[string]any {
	type group struct {
		name    string
		counts  counts
		vectors []map[string]any
	}
	groupIdx := map[string]int{}
	var groups []*group
	var totals counts
	var totalTime time.Duration

	for res := range results {
		if meta.OmitSkipped && res.Outcome == s3tests.Skipped {
			continue
		}
		i, ok := groupIdx[res.Group]
		if !ok {
			i = len(groups)
			groupIdx[res.Group] = i
			groups = append(groups, &group{name: res.Group})
		}
		g := groups[i]
		g.counts.add(res)
		totals.add(res)
		totalTime += res.Duration
		g.vectors = append(g.vectors, vectorView(res))
	}

	sort.Slice(groups, func(i, j int) bool { return groups[i].name < groups[j].name })

	firstFail := ""
	groupViews := make([]map[string]any, 0, len(groups))
	for _, g := range groups {
		sort.Slice(g.vectors, func(i, j int) bool {
			return g.vectors[i]["id"].(string) < g.vectors[j]["id"].(string)
		})
		if firstFail == "" {
			for _, vec := range g.vectors {
				if vec["badge"] == "fail" {
					firstFail = vec["id"].(string)
					break
				}
			}
		}
		groupViews = append(groupViews, map[string]any{
			"name":     g.name,
			"pctClass": g.counts.pctClass(),
			"barWidth": g.counts.barWidth(),
			"open":     g.counts.Fail+g.counts.Errors > 0,
			"counts":   g.counts.view(),
			"vectors":  g.vectors,
		})
	}

	generated := ""
	if !meta.GeneratedAt.IsZero() {
		generated = meta.GeneratedAt.UTC().Format("2006-01-02 15:04:05 UTC")
	}
	return map[string]any{
		"target":           meta.Target,
		"hasTarget":        meta.Target != "",
		"corpusVersion":    meta.CorpusVersion,
		"hasCorpusVersion": meta.CorpusVersion != "",
		"generated":        generated,
		"hasGenerated":     generated != "",
		"totalTime":        humanDuration(totalTime),
		"properties":       sortedProperties(meta.Properties),
		"hasProvenance":    meta.Target != "" || meta.CorpusVersion != "" || generated != "" || len(meta.Properties) > 0,
		"totals":           totals.view(),
		"firstFail":        firstFail,
		"groups":           groupViews,
	}
}

type counts struct {
	Pass, Fail, Blocked, Errors, Skipped, Total int
}

func (c *counts) add(res s3tests.VectorResult) {
	c.Total++
	switch res.Outcome {
	case s3tests.Pass:
		c.Pass++
	case s3tests.Fail:
		if res.RunnerError != "" {
			c.Errors++
		} else {
			c.Fail++
		}
	case s3tests.Blocked:
		c.Blocked++
	case s3tests.Skipped:
		c.Skipped++
	}
}

// attempted is the pass-rate denominator: everything except skipped.
func (c *counts) attempted() int { return c.Total - c.Skipped }

func (c *counts) view() map[string]any {
	return map[string]any{
		"pass": c.Pass, "fail": c.Fail, "blocked": c.Blocked,
		"errors": c.Errors, "skipped": c.Skipped, "total": c.Total,
		"attempted": c.attempted(),
		"passPct":   c.passPct(),
		"hasFail":   c.Fail > 0, "hasBlocked": c.Blocked > 0,
		"hasErrors": c.Errors > 0, "hasSkipped": c.Skipped > 0,
		"failZero": c.Fail == 0, "blockedZero": c.Blocked == 0,
		"errorsZero": c.Errors == 0, "skippedZero": c.Skipped == 0,
	}
}

// passPct renders the pass rate ("98.2%", one decimal, round half away from
// zero via integer arithmetic — the shared cross-implementation rule), or a
// dash when nothing was attempted.
func (c *counts) passPct() string {
	attempted := c.attempted()
	if attempted == 0 {
		return "—"
	}
	p10 := (1000*c.Pass + attempted/2) / attempted
	return fmt.Sprintf("%d.%d%%", p10/10, p10%10)
}

// pctClass maps the pass rate onto the coverage-style watermarks using exact
// integer comparisons (high >= 80%, medium >= 50%).
func (c *counts) pctClass() string {
	attempted := c.attempted()
	switch {
	case attempted == 0:
		return ""
	case 100*c.Pass >= 80*attempted:
		return "high"
	case 100*c.Pass >= 50*attempted:
		return "medium"
	default:
		return "low"
	}
}

// barWidth is the pass-fraction bar fill width in whole percent.
func (c *counts) barWidth() string {
	attempted := c.attempted()
	if attempted == 0 {
		return "0"
	}
	return fmt.Sprintf("%d", (100*c.Pass+attempted/2)/attempted)
}

// definitionBase is the raw-file location of the corpus vector files; a
// text-fragment URL on it opens the file scrolled to the vector's "id" line.
const definitionBase = "https://raw.githubusercontent.com/cloud-portable/s3vectors/main/vectors/"

// definitionURL links a vector card to the vector's definition in the corpus
// repository. Group and id are plain concatenated, not percent-encoded: the
// corpus schema restricts both to [a-z0-9-], and avoiding an encoder keeps
// the Go and JS reporters byte-identical (their encoders escape different
// characters). The fragment prefix is the pre-encoded form of `"id": "`.
func definitionURL(group, id string) string {
	return definitionBase + group + ".json#:~:text=%22id%22%3A%20%22" + id + "%22"
}

func vectorView(res s3tests.VectorResult) map[string]any {
	badge := string(res.Outcome)
	reason, summary, detail := "", "", ""
	switch res.Outcome {
	case s3tests.Fail:
		summary, detail = failureDetail(res)
		if res.RunnerError != "" {
			badge = "error"
			summary = "runner error: " + res.RunnerError
			if detail == res.RunnerError {
				detail = "" // nothing beyond what the message already says
			}
		}
	case s3tests.Blocked:
		reason = "blocked: " + res.Reason
	case s3tests.Skipped:
		reason = "skipped: " + res.Reason
	}
	tags := res.Tags
	if tags == nil {
		tags = []string{}
	}
	warnings := res.Warnings
	if warnings == nil {
		warnings = []string{}
	}
	return map[string]any{
		"id": res.ID, "badge": badge,
		"duration": vectorDuration(res.Duration),
		"title":    res.Title, "hasTitle": res.Title != "",
		"tags": tags, "hasTags": len(tags) > 0,
		"reason": reason, "hasReason": reason != "",
		"summary": summary, "hasSummary": summary != "",
		"detail": detail, "hasDetail": detail != "",
		"warnings": warnings,
		"source":   res.Source, "hasSource": res.Source != "",
		"definitionURL": definitionURL(res.Group, res.ID),
		"hasDesc":       res.Title != "" || len(tags) > 0,
		"hasOutcome":    reason != "" || summary != "" || detail != "" || len(warnings) > 0,
	}
}

// failureDetail summarizes the failing (last executed) step. The summary
// names only the step — the mismatches live solely in the detail block, so
// nothing renders twice.
func failureDetail(res s3tests.VectorResult) (summary, detail string) {
	if len(res.Steps) == 0 {
		return res.Reason, ""
	}
	step := res.Steps[len(res.Steps)-1]
	summary = fmt.Sprintf("step %d (%s) failed", step.Index+1, step.Name)
	var lines []string
	if step.Err != "" {
		lines = append(lines, step.Err)
	}
	for _, f := range step.Failures {
		lines = append(lines, fmt.Sprintf("%s: expected %s, got %s", f.Field, f.Expected, f.Actual))
	}
	return summary, strings.Join(lines, "\n")
}

// vectorDuration renders a per-vector duration as seconds with exactly three
// decimals, from whole milliseconds (round half away from zero) — the shared
// cross-implementation rule.
func vectorDuration(d time.Duration) string {
	ms := (d.Nanoseconds() + 500_000) / 1_000_000
	return fmt.Sprintf("%d.%03ds", ms/1000, ms%1000)
}

// humanDuration renders a duration compactly: "42.3s" under a minute
// (one decimal, round half away from zero), Go-style "4m12s" above (rounded
// to whole seconds).
func humanDuration(d time.Duration) string {
	if d < time.Minute {
		ds := (d.Nanoseconds() + 50_000_000) / 100_000_000
		return fmt.Sprintf("%d.%ds", ds/10, ds%10)
	}
	return d.Round(time.Second).String()
}

func sortedProperties(props map[string]string) []map[string]string {
	keys := make([]string, 0, len(props))
	for k := range props {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	out := make([]map[string]string, 0, len(keys))
	for _, k := range keys {
		out = append(out, map[string]string{"name": k, "value": props[k]})
	}
	return out
}

File: packages/python/src/cloud_portable_s3tests/report/html.py
"""HTML reporter rendering the shared cross-implementation mustache template
(``page.mustache`` — a synced copy of ``shared/report/page.mustache``). The
view model and all number formatting follow the shared contract so identical
results render byte-identical pages in every language, enforced by the shared
golden test (``shared/report/fixture.json`` + ``golden.html``; the Go
implementation is canonical)."""

from __future__ import annotations

from dataclasses import dataclass
from importlib import resources
from typing import Any, BinaryIO, Iterable

from .._result import VectorResult
from .._timefmt import secs1, secs3, utc
from . import Meta
from ._mustache import Template

_template = Template(resources.files(__package__).joinpath("page.mustache").read_text("utf-8"))


def write(w: BinaryIO, results: Iterable[VectorResult], meta: Meta | None = None) -> None:
    """Consume the results and write the report page (UTF-8) to ``w``."""
    w.write(render(results, meta).encode("utf-8"))


def render(results: Iterable[VectorResult], meta: Meta | None = None) -> str:
    return _template.render(view_model(results, meta or Meta()))


@dataclass
class _Counts:
    pass_: int = 0
    fail: int = 0
    blocked: int = 0
    errors: int = 0
    skipped: int = 0
    total: int = 0

    def add(self, res: VectorResult) -> None:
        self.total += 1
        if res.outcome == "pass":
            self.pass_ += 1
        elif res.outcome == "fail":
            if res.runner_error:
                self.errors += 1
            else:
                self.fail += 1
        elif res.outcome == "blocked":
            self.blocked += 1
        elif res.outcome == "skipped":
            self.skipped += 1

    @property
    def attempted(self) -> int:
        return self.total - self.skipped

    def view(self) -> dict[str, Any]:
        return {
            "pass": self.pass_,
            "fail": self.fail,
            "blocked": self.blocked,
            "errors": self.errors,
            "skipped": self.skipped,
            "total": self.total,
            "attempted": self.attempted,
            "passPct": self.pass_pct(),
            "hasFail": self.fail > 0,
            "hasBlocked": self.blocked > 0,
            "hasErrors": self.errors > 0,
            "hasSkipped": self.skipped > 0,
            "failZero": self.fail == 0,
            "blockedZero": self.blocked == 0,
            "errorsZero": self.errors == 0,
            "skippedZero": self.skipped == 0,
        }

    # Shared integer-arithmetic formatting rules (see the Go implementation).
    def pass_pct(self) -> str:
        a = self.attempted
        if a == 0:
            return "—"
        p10 = (1000 * self.pass_ + a // 2) // a
        return f"{p10 // 10}.{p10 % 10}%"

    def pct_class(self) -> str:
        a = self.attempted
        if a == 0:
            return ""
        if 100 * self.pass_ >= 80 * a:
            return "high"
        if 100 * self.pass_ >= 50 * a:
            return "medium"
        return "low"

    def bar_width(self) -> str:
        a = self.attempted
        if a == 0:
            return "0"
        return str((100 * self.pass_ + a // 2) // a)


def view_model(results: Iterable[VectorResult], meta: Meta) -> dict[str, Any]:
    groups: dict[str, dict[str, Any]] = {}
    totals = _Counts()
    total_time = 0
    for res in results:
        if meta.omit_skipped and res.outcome == "skipped":
            continue
        g = groups.get(res.group)
        if g is None:
            g = {"name": res.group, "counts": _Counts(), "vectors": []}
            groups[res.group] = g
        g["counts"].add(res)
        totals.add(res)
        total_time += res.duration
        g["vectors"].append(vector_view(res))

    first_fail = ""
    group_views = []
    for name in sorted(groups):
        g = groups[name]
        vectors = sorted(g["vectors"], key=lambda v: v["id"])
        if first_fail == "":
            f = next((v for v in vectors if v["badge"] == "fail"), None)
            if f is not None:
                first_fail = f["id"]
        c: _Counts = g["counts"]
        group_views.append(
            {
                "name": name,
                "pctClass": c.pct_class(),
                "barWidth": c.bar_width(),
                "open": c.fail + c.errors > 0,
                "counts": c.view(),
                "vectors": vectors,
            }
        )

    generated = format_generated(meta.generated_at) if meta.generated_at is not None else ""
    properties = [{"name": k, "value": meta.properties[k]} for k in sorted(meta.properties)]
    return {
        "target": meta.target,
        "hasTarget": meta.target != "",
        "corpusVersion": meta.corpus_version,
        "hasCorpusVersion": meta.corpus_version != "",
        "generated": generated,
        "hasGenerated": generated != "",
        "totalTime": human_duration(total_time),
        "properties": properties,
        "hasProvenance": meta.target != "" or meta.corpus_version != "" or generated != "" or len(properties) > 0,
        "totals": totals.view(),
        "firstFail": first_fail,
        "groups": group_views,
    }


# Raw-file location of the corpus vector files; a text-fragment URL on it
# opens the file scrolled to the vector's "id" line.
DEFINITION_BASE = "https://raw.githubusercontent.com/cloud-portable/s3vectors/main/vectors/"


def definition_url(group: str, id: str) -> str:
    """Link to the vector's definition in the corpus repository. Group and id
    are plain concatenated, not percent-encoded: the corpus schema restricts
    both to [a-z0-9-], and avoiding an encoder keeps every reporter
    byte-identical. The fragment prefix is the pre-encoded form of `"id": "`."""
    return f"{DEFINITION_BASE}{group}.json#:~:text=%22id%22%3A%20%22{id}%22"


def vector_view(res: VectorResult) -> dict[str, Any]:
    badge = res.outcome
    reason = summary = detail = ""
    if res.outcome == "fail":
        summary, detail = failure_detail(res)
        if res.runner_error:
            badge = "error"
            summary = "runner error: " + res.runner_error
            if detail == res.runner_error:
                detail = ""  # nothing beyond what the message already says
    elif res.outcome == "blocked":
        reason = "blocked: " + res.reason
    elif res.outcome == "skipped":
        reason = "skipped: " + res.reason
    tags = list(res.tags)
    warnings = list(res.warnings)
    return {
        "id": res.id,
        "badge": badge,
        "duration": vector_duration(res.duration),
        "title": res.title,
        "hasTitle": res.title != "",
        "tags": tags,
        "hasTags": len(tags) > 0,
        "reason": reason,
        "hasReason": reason != "",
        "summary": summary,
        "hasSummary": summary != "",
        "detail": detail,
        "hasDetail": detail != "",
        "warnings": warnings,
        "source": res.source,
        "hasSource": res.source != "",
        "definitionURL": definition_url(res.group, res.id),
        "hasDesc": res.title != "" or len(tags) > 0,
        "hasOutcome": reason != "" or summary != "" or detail != "" or len(warnings) > 0,
    }


def failure_detail(res: VectorResult) -> tuple[str, str]:
    """Summarize the failing (last executed) step. The summary names only the
    step — the mismatches live solely in the detail block."""
    if not res.steps:
        return res.reason, ""
    step = res.steps[-1]
    summary = f"step {step.index + 1} ({step.name}) failed"
    lines = []
    if step.err:
        lines.append(step.err)
    for f in step.failures:
        lines.append(f"{f.field}: expected {f.expected}, got {f.actual}")
    return summary, "\n".join(lines)


def vector_duration(ns: int) -> str:
    """Seconds with exactly three decimals from whole ms (round half up)."""
    return secs3(ns) + "s"


def human_duration(ns: int) -> str:
    """"42.3s" under a minute; Go's Duration.String() form above, rounded to
    whole seconds ("4m12s", "1h0m0s")."""
    if ns < 60_000_000_000:
        return secs1(ns) + "s"
    secs = (ns + 500_000_000) // 1_000_000_000
    h, secs = divmod(secs, 3600)
    m, secs = divmod(secs, 60)
    return f"{h}h{m}m{secs}s" if h > 0 else f"{m}m{secs}s"


def format_generated(d) -> str:
    d = utc(d)
    return f"{d.year:04d}-{d.month:02d}-{d.day:02d} {d.hour:02d}:{d.minute:02d}:{d.second:02d} UTC"
